pyrometheus/__init__.py

```diff
@@ -329,37 +330,84 @@ def production_rate_expr(sol: ct.Solution, species, r_net):
               for prod_index in indices_rev]
     sum_fwd = sum(nu*r_net[index] for nu, index in zip(nu_fwd, indices_fwd))
     sum_rev = sum(nu*r_net[index] for nu, index in zip(nu_rev, indices_rev))
-    return sum_rev - sum_fwd
+    return (sum_rev - sum_fwd) * ones
 
 # }}}
 
 
 # {{{ main code template
 
 code_tpl = Template(
-    """
-import numpy as np
+    """import numpy as np
+
+
+def _pyro_make_array(res_list):
+    # https://github.com/numpy/numpy/issues/18004
+
+    from numbers import Number
+    all_numbers = all(isinstance(e, Number) for e in res_list)
+
+    dtype = np.float64 if all_numbers else np.object
+    result = np.empty((len(res_list),), dtype=dtype)
+
+    # 'result[:] = res_list' may look tempting, however:
+    # https://github.com/numpy/numpy/issues/16564
+    for idx in range(len(res_list)):
+        result[idx] = res_list[idx]
+
+    return result
+
+
+def _pyro_norm(actx_np, argument, normord):
+    # Wrap norm for scalars
+    from numbers import Number
+    if isinstance(argument, Number):
+        return np.abs(argument)
+    return actx_np.linalg.norm(argument, normord)
+
 
 class Thermochemistry:
-    model_name    = ${repr(sol.source)}
-    num_elements  = ${sol.n_elements}
-    num_species   = ${sol.n_species}
-    num_reactions = ${sol.n_reactions}
-    num_falloff   = ${
-        sum(1 if isinstance(r, ct.FalloffReaction) else 0
+    def __init__(self, actx_np=np):
+        \"""Initialize the mechanism object.
+
+        Parameters
+        ----------
+        actx_np
+            numpy-like namespace providing at least the following functions:
+            actx_np.log(T)
+            actx_np.log10(T)
+            actx_np.exp(T)
+            actx_np.where(T > 0, T_yes, T_no)
+            actx_np.linalg.norm(T, np.inf)
+            where T is model user data having the same size, shape, and nature
+            of the user's temperature data, for example. This parameter allows
+            the user to provide a numpy work-alike that is capable of processing
+            the user's data structures, or even invoking custom kernels for the
+            associated operations. This parameter defaults to *actual numpy*, so
+            unless it is needed by the user (e.g. for GPU processing).
```

Comment:
This is a good start to document what is expect of `actx_np`, but it needs to be pulled out into a separate interface and actually make functional use of RST markup.

Reply:
Plz see below